src/main/java/org/team1540/robot2025/subsystems/drive/OdometryThread.java

```diff
@@ -0,0 +1,150 @@
+package org.team1540.robot2025.subsystems.drive;
+
+import com.ctre.phoenix6.BaseStatusSignal;
+import com.ctre.phoenix6.StatusSignal;
+import edu.wpi.first.units.measure.Angle;
+import edu.wpi.first.wpilibj.RobotController;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.DoubleSupplier;
+import org.team1540.robot2025.generated.TunerConstants;
+
+/**
+ * Provides an interface for asynchronously reading high-frequency measurements to a set of queues.
+ *
+ * <p>This version is intended for Phoenix 6 devices on both the RIO and CANivore buses. When using
+ * a CANivore, the thread uses the "waitForAll" blocking method to enable more consistent sampling.
+ * This also allows Phoenix Pro users to benefit from lower latency between devices using CANivore
+ * time synchronization.
+ */
+public class OdometryThread extends Thread {
+    private final Lock signalsLock = new ReentrantLock(); // Prevents conflicts when registering signals
+    private BaseStatusSignal[] phoenixSignals = new BaseStatusSignal[0];
+    private final List<DoubleSupplier> genericSignals = new ArrayList<>();
+    private final List<Queue<Double>> phoenixQueues = new ArrayList<>();
+    private final List<Queue<Double>> genericQueues = new ArrayList<>();
+    private final List<Queue<Double>> timestampQueues = new ArrayList<>();
+
+    private static final boolean isCANFD = TunerConstants.kCANBus.isNetworkFD();
+    private static OdometryThread instance = null;
+
+    public static OdometryThread getInstance() {
+        if (instance == null) {
+            instance = new OdometryThread();
+        }
+        return instance;
+    }
+
+    private OdometryThread() {
+        setName("OdometryThread");
+        setDaemon(true);
+    }
+
+    @Override
+    public void start() {
+        if (!timestampQueues.isEmpty()) {
+            super.start();
+        }
+    }
+
+    /** Registers a Phoenix signal to be read from the thread. */
+    public Queue<Double> registerSignal(StatusSignal<Angle> signal) {
+        Queue<Double> queue = new ArrayBlockingQueue<>(20);
+        signalsLock.lock();
+        Drivetrain.odometryLock.lock();
+        try {
+            BaseStatusSignal[] newSignals = new BaseStatusSignal[phoenixSignals.length + 1];
+            System.arraycopy(phoenixSignals, 0, newSignals, 0, phoenixSignals.length);
+            newSignals[phoenixSignals.length] = signal;
+            phoenixSignals = newSignals;
```

Comment:
linked list

Reply:
1) there aren't enough adds here for that to be worth it
2) you iterate over the array more often than you add to it, and iterating over an array is faster than iterating over a linked list because of memory locality, since you can load a good chunk of an array into the cpu cache in one go, which isn't possible with a linked list